application/controllers/UploaderController.php

```diff
@@ -141,8 +141,9 @@ public function run($actionID)
                 mkdir($sTempUploadDir);
             }
             $filename = $_FILES['uploadfile']['name'];
-            // Do we filter file name ? It's used on displaying only , but not save like that.
-            //$filename = sanitize_filename($_FILES['uploadfile']['name']);// This remove all non alpha numeric characters and replaced by _ . Leave only one dot .
+            /* Fix file name : need to get only the basename url encoded for display . Avoid usage of basename : need valid UTF8 setlocale */
+            $aFilename = explode(DIRECTORY_SEPARATOR, rtrim( $filename, DIRECTORY_SEPARATOR));
```

Comment:
I don't think that is safe.
See https://www.php.net/manual/en/features.file-upload.post-method.php
"$_FILES['userfile']['name'] : The original name of the file on the client machine."
Why would the server DIRECTORY_SEPARATOR matter here?

Reply:
To replace basename($filename) only. 
I don't know if we need it and when it was added.

Reply:
Then it is a bad replacement. 
$_FILES['userfile']['name'] is not to be trusted at all, since it is user content. If you want to trim it of any dangerous content you will have to consider all kind of directory separators and also consider other characters (dots, quotes. etc)

Reply:
basename($filename) do only that : no other filter except last part.

I can use https://www.php.net/manual/fr/function.basename.php#121405 if you think it's more secure. Then if a linux user send bad\name.png : we save only name.png

Else: without valid setlocale : i don't have any solution :), except remove any filter on name (we don't really care on tghe server).

Reply:
I would use sanitize_filename(), not sure if it also handles file paths - please check.

Reply:
> I would use sanitize_filename(), not sure if it also handles file paths - please check.

[sanitize_filename](https://github.com/LimeSurvey/LimeSurvey/blob/f98ca8b9697e819449a4d7c79f224ca8ce4b2b19/application/helpers/sanitize_helper.php#L119) broke with same issue : österreich.png > sterreich.png 
with `setlocale(LC_CTYPE,"C");`
 maybe usage of pathinfo ?

Reply:
Oups … bad host for testing …

Reply:
No : confirmed : our current sanitize_filename broke with same issue : österreich.png > sterreich.png

Reply:
PS : even if fix the way of saving : https://github.com/LimeSurvey/LimeSurvey/blob/f98ca8b9697e819449a4d7c79f224ca8ce4b2b19/application/controllers/ResponsesController.php#L645

We sanitize here : maybe lets close issue as "Server configuration issue" ?

Reply:
Hmm.. can you try to figure out what function removes the umlaut in sanitize_filename ?

Reply:
I try ,

Else : i don't like current sanitizer : didn't accept `[``]` , but why ? There are no reason here ?

Reply:
I'd very much advise against making this a configuration issue. LANG=C is the default on all Debian based Apache installations, so the majority of installations would have a configuration issue:
https://salsa.debian.org/apache-team/apache2/-/blob/master/debian/config-dir/envvars
That sounds unrealistic.

Reply:
Related issue : 17718: sanitize_filename didn't really fix filename for all system

Reply:
> Hmm.. can you try to figure out what function removes the umlaut in sanitize_filename ?

for sanitize_filename : `mb_strcut` usage remove the 1st utf8 character (with C locale), 
https://github.com/LimeSurvey/LimeSurvey/blob/b43948123d3cb6e1a029779df0b346a704af3a5a/application/helpers/sanitize_helper.php#L143